libs/arcade-tdk/arcade_tdk/model.py

```diff
@@ -0,0 +1,155 @@
+
+        try:
+            secret_value = context.get_secret(secret_key)
+        except ValueError:
+            pass
+        else:
+            try:
+                max_concurrent_requests = int(secret_value)
+            except ValueError:
+                raise ValueError(
+                    f"Invalid value set for the secret '{secret_key}'. "
+                    f"Expected a numeric string, got '{secret_value}'."
+                )
+
+        if max_concurrent_requests < 1:
+            raise ValueError(
+                "Invalid value set for 'max_concurrent_requests'. "
+                f"Expected a positive integer, got '{max_concurrent_requests}'."
+            )
+
+        self.max_concurrent_requests = max_concurrent_requests
+        self._semaphore = asyncio.Semaphore(self.max_concurrent_requests)
+
+    @property
+    @abstractmethod
+    def service_name(self) -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def default_api_version(self) -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def default_headers(self) -> dict[str, str]:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def base_url(self) -> str:
+        raise NotImplementedError
+
+    @property
+    def auth_token(self) -> str | None:
+        return self.context.get_auth_token_or_empty()
+
+    @abstractmethod
+    def _build_url(self, path: str) -> str:
+        raise NotImplementedError
+
+    async def request(
+        self,
+        method: str,
+        path: str,
+        headers: dict[str, str] | None = None,
+        **kwargs,
+    ) -> dict:
+        headers = {**self.default_headers, **(headers or {})}
+        url = self._build_url(path=path)
+        async with self._semaphore, httpx.AsyncClient() as client:
+            response = await client.request(method, url, headers=headers, **kwargs)
+            self._raise_for_status(response)
+            return self._format_response_dict(response)
+
+    async def get(self, path: str, headers: dict[str, str] | None = None, **kwargs) -> dict:
+        return await self.request("GET", path, headers=headers, **kwargs)
+
+    async def post(self, path: str, headers: dict[str, str] | None = None, **kwargs) -> dict:
+        return await self.request("POST", path, headers=headers, **kwargs)
+
+    async def put(self, path: str, headers: dict[str, str] | None = None, **kwargs) -> dict:
+        return await self.request("PUT", path, headers=headers, **kwargs)
+
+    async def delete(self, path: str, headers: dict[str, str] | None = None, **kwargs) -> dict:
+        return await self.request("DELETE", path, headers=headers, **kwargs)
+
+    def _format_response_dict(self, response: httpx.Response) -> dict:
+        try:
+            return cast(dict, response.json())
+        except (UnicodeDecodeError, json.decoder.JSONDecodeError):
+            return {"response": response.text}
```

Comment:
opinion: I like the idea of always returning the same type by default, but I also think the tool that is calling this client should have the freedom to disable this. Since this will be a base class for many many many integrations I think flexibility is paramount.